### PR TITLE
cray platform: detect platform properly after module purge

### PR DESCRIPTION
**Pull request by becker33**
Currently, we assume the cray platform can be identified by the default modules

We have discovered a cray machine on which the default modules are `No modules loaded`.

This caused some problems for Spack's platform detection and OS setup. This PR allows Spack to persevere despite those problems. In the worst case in which the `cle_release` files have been removed from the system, we will continue with `os=cnlunknown` instead of failing immediately.

NOTE 2019-10-01: The scope of this PR has expanded to include a set of hotfixes found/used at an internal hackathon at LLNL.

### CODE REVIEW

File: lib/spack/spack/platforms/cray.py [hunk 1]
```diff
@@ -106,7 +114,7 @@ def setup_platform_environment(cls, pkg, env):
 
     @classmethod
     def detect(cls):
-        return os.environ.get('CRAYPE_VERSION') is not None
+        return 'CRAYPE_VERSION' in os.environ or os.path.exists('/opt/cray/pe')
```
Comment (alalazo) [inline lib/spack/spack/platforms/cray.py]: This change seems to be fine, in the sense that it returns `True` even after a `module purge`. I am wondering if this can be further simplified to:
```suggestion
        return os.path.exists('/opt/cray/pe')
```
but I am not sure of it as I have currently access only to Cori.

Comment (alalazo) [reply lib/spack/spack/platforms/cray.py]: On Cori after a `module purge` the only `CRAY_*` variable available is:
```console
culpo@cori10:~/github/spack> echo $CRAY_SITE_LIST_DIR
/etc/opt/cray/pe/modules
```
so maybe this might also be a good candidate to check if platform is `cray`? To be clear, I'm not suggesting a change here, just noting that a variable stays there after we pruned all the modules.

File: lib/spack/spack/platforms/cray.py [hunk 2]
```diff
@@ -47,31 +62,24 @@ def __init__(self):
...
-                self.add_target(name, self.targets[safe_name])
-
-        if self.back_end is not None:
-            self.default = self.back_end
-            self.add_target('default', self.targets[self.back_end])
-        else:
+        self.back_end = os.environ.get('SPACK_BACK_END',
+                                       self._default_target_from_env())
+        self.default = self.back_end
+        if self.back_end not in self.targets:
```
Comment (alalazo) [inline lib/spack/spack/platforms/cray.py:55]: It seems that after a `module purge` we might end up with an empty set of targets, so this check is always `True`.

File: lib/spack/spack/platforms/cray.py [hunk 3]
```diff
@@ -16,20 +18,33 @@
...
             if len(i.split()) == 1]
 
 
-def _fill_craype_targets_from_modules(targets, modules):
+def _fill_craype_target_names_from_modules(targets, modules):
     '''Extend CrayPE CPU targets list with those found in list of modules.'''
     # Craype- module prefixes that are not valid CPU targets.
-    non_targets = ('hugepages', 'network', 'target', 'accel', 'xtpe')
+    non_targets = (
+        'hugepages', 'network', 'target', 'accel', 'xtpe', 'dl-plugin')
```
Comment (alalazo) [inline lib/spack/spack/platforms/cray.py]: This also seems to be fine, and unit testable. I'll change it to actually return the list so that it does not need to be passed as an argument.

File: lib/spack/spack/platforms/cray.py [hunk 4]
```diff
@@ -118,17 +126,28 @@ def _default_target_from_env(self):
...
+                    return platform.machine()
 
     def _avail_targets(self):
         '''Return a list of available CrayPE CPU targets.'''
         if getattr(self, '_craype_targets', None) is None:
             output = module('avail', '-t', 'craype-')
             craype_modules = _get_modules_in_modulecmd_output(output)
             self._craype_targets = targets = []
-            _fill_craype_targets_from_modules(targets, craype_modules)
+            _fill_craype_target_names_from_modules(targets, craype_modules)
```
Comment (alalazo) [inline lib/spack/spack/platforms/cray.py]: Above we assume that `module -t av craype-` will return something sensible, but it might not be the case after a `module purge`. Is it worth as a backup option to look at the name of files in locations like:
```
/opt/cray/pe/craype/X.Y.Z/modulefiles
```
or would it be a problem for any reason (too much stress on the filesystem / location changes depending on Cray machine, etc. )?

Also, would it be worth in the long term to cache somewhere on file the result of this detection and refresh the cache every so often (or on demand)?

File: lib/spack/spack/platforms/cray.py (hunk 1 above)
Comment (becker33) [reply lib/spack/spack/platforms/cray.py]: That variable is not set on the machine I've been testing on.

File: lib/spack/spack/platforms/cray.py (hunk 2 above)
Comment (becker33) [reply lib/spack/spack/platforms/cray.py:55]: We check `module avail`, not `module list`, to get `self._avail_targets`

File: lib/spack/spack/platforms/cray.py (hunk 4 above)
Comment (becker33) [reply lib/spack/spack/platforms/cray.py]: I would lean away from a file cache because we could be operating out of a directory cross-mounted on two machines.

File: lib/spack/spack/platforms/cray.py (hunk 1 above)
Comment (becker33) [reply lib/spack/spack/platforms/cray.py]: Should the check really be `'/opt/cray' in os.environ.get('MODULEPATH', '')`? I think the fundamental quality of a cray platform is the availability of the PE through modules.

File: lib/spack/spack/platforms/cray.py (hunk 2 above)
Comment (alalazo) [reply lib/spack/spack/platforms/cray.py:55]: It is my impression that purging some modules, like `craype-*`, removes a few paths from `MODULEPATH`, so there will be modules not available after the purge.

File: lib/spack/spack/operating_systems/cray_backend.py [hunk 5]
```diff
@@ -79,7 +79,10 @@ class Cnl(OperatingSystem):
     def __init__(self):
         name = 'cnl'
         version = self._detect_crayos_version()
-        super(Cnl, self).__init__(name, version)
+        if version:
+            super(LinuxDistro, self).__init__(name, version)
```
Comment (alalazo) [inline lib/spack/spack/operating_systems/cray_backend.py:88]: That's nasty, here we are skipping one step in the MRO :slightly_smiling_face:

Comment (becker33) [reply lib/spack/spack/operating_systems/cray_backend.py:88]: I couldn't see a way around it to accomplish the same goal... python lets you do evil things to MRO

File: lib/spack/spack/platforms/cray.py (hunk 2 above)
Comment (becker33) [reply lib/spack/spack/platforms/cray.py:55]: I think that's true, and that the machine I'm testing on isn't quite the same as after a module purge. It's like after a module purge but with those paths added back in.

Anyway, your PR to this branch fixes that

File: lib/spack/spack/test/operating_system.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
-import spack.operating_systems.cnl as cnl
+import spack.operating_systems.cray_backend as crauy_backend
```
Comment (alalazo) [inline lib/spack/spack/test/operating_system.py]: Typo above

File: lib/spack/spack/operating_systems/cray_backend.py (hunk 5 above)
Comment (dev-zero) [reply lib/spack/spack/operating_systems/cray_backend.py:88]: this seems like a good application for an intermediate Factory to return a class with the proper mixin

File: lib/spack/spack/operating_systems/cray_backend.py [hunk 7]
```diff
@@ -95,8 +98,7 @@ def _detect_crayos_version(cls):
             v = read_clerelease_file()
             return spack.version.Version(v)[0]
         else:
-            raise spack.error.UnsupportedPlatformError(
-                'Unable to detect Cray OS version')
+            return None
```
Comment (dev-zero) [inline lib/spack/spack/operating_systems/cray_backend.py:114]: this may lead to hard to debug errors if that ever crops up somewhere, what is the reason this can't raise an error?

Comment (becker33) [reply lib/spack/spack/operating_systems/cray_backend.py:114]: Because not all Cray systems run CNL on the backend. We have a system at Livermore running in what Cray calls "cluster" mode that runs RHEL7 under the Cray PE.

So if we don't detect any Cray OS version on the system, we return `None`. We can't ever be sure we will get a Cray OS version. I thought this was cleaner than a `try/except` block in the client code.

Comment (dev-zero) [reply lib/spack/spack/operating_systems/cray_backend.py:114]: given that exceptions are first-class citizens in Python, personally I would rather ensure that `__str__` returns a string and raises an exception if it is unable to do so

File: lib/spack/spack/operating_systems/cray_backend.py (hunk 5 above)
Comment (alalazo) [reply lib/spack/spack/operating_systems/cray_backend.py:88]: @dev-zero The intermediate factory is a meta class in Python and might be an overkill here.

**Comment by chuckatkins** [comment]
The CCE changes look like they're based on the assumption that v9.0 and greater are using the new clang interface.  This isn't necessarily the case as the the `cce-legacy` module can be used that still preserves the older interface with newer versions of cce.  So far in CMake we've been able to treat the new cce interface as proper clang and not make the distinction between clang and cray-clang.  That has not been the case for other compilers like XL where we explicitly have both `XL` and `XLClang` compiler id's, but so far that hasn't been an issue for cce and we can just have `Cray` for the older legacy interface and have the new one use the same `Clang` interface as vanilla clang.

**Comment by chuckatkins** [comment]
I suppose this goes back to a previous discussion we had regarding the cray detection.  My question would be whether or not spack even should be detecting the cray platform if no modules are loaded?  If no modules are loaded then should the machine not be treated as vanilla linux for the front end?

File: lib/spack/spack/compilers/cce.py [hunk 8]
```diff
@@ -60,6 +71,8 @@ def c99_flag(self):
 
     @property
     def c11_flag(self):
+        if self.verions >= ver('9.0'):
```
Comment (dev-zero) [inline lib/spack/spack/compilers/cce.py]: typo: `self.verions` -> `self.version`

File: lib/spack/spack/operating_systems/cray_backend.py (hunk 7 above)
Comment (becker33) [reply lib/spack/spack/operating_systems/cray_backend.py:114]: `str` already ensures that it returns a string, this is in `_detect_crayos_version`.

File: lib/spack/spack/compiler.py [hunk 9]
```diff
@@ -406,6 +406,40 @@ def c11_flag(self):
                                       "the C11 standard",
                                       "c11_flag")
 
+    #
+    # Recompute the compiler version for a compiler
+    # Take advantage of modules to be cross-platform
+    # Not a class method
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: why does this need to say "not a class method"?

File: lib/spack/spack/compiler.py [hunk 10]
```diff
@@ -406,6 +406,40 @@ def c11_flag(self):
                                       "the C11 standard",
                                       "c11_flag")
 
+    #
+    # Recompute the compiler version for a compiler
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: This needs to be more specific than "recompute" -- does it run the compiler? Does it just ask the compiler or does it actually "compute" something?

File: lib/spack/spack/compiler.py [hunk 11]
```diff
@@ -406,6 +406,40 @@ def c11_flag(self):
...
+    def compute_real_version(self):
+        """Recompute the compiler version for a compiler.
+
+        This method takes advantage of compiler modules to be cross-platform.
+        """
+        # store environment to replace later
+        backup_env = os.environ
+
+        # load modules and set env variables
+        # See build_environment.py for comments on this code
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: there should really be relevant comments right here, with "see build_environment.py for more details".

File: lib/spack/spack/compiler.py [hunk 12]
```diff
@@ -406,6 +406,40 @@ def c11_flag(self):
...
                                       "c11_flag")
 
+    #
+    # Recompute the compiler version for a compiler
+    # Take advantage of modules to be cross-platform
+    # Not a class method
+    def compute_real_version(self):
+        """Recompute the compiler version for a compiler.
+
+        This method takes advantage of compiler modules to be cross-platform.
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: "takes advantage" and "compiler modules" are vague here.  Are "compiler modules" environment modules?  No, right?  What does "takes advantage of" mean?  Make this more specific.

Also what does "real" mean? I'm assuming that means we ask the compiler instead of trusting what is in `compilers.yaml` but there isn't enough info in the comment for me to know that.

File: lib/spack/spack/compiler.py [hunk 13]
```diff
@@ -406,6 +406,40 @@ def c11_flag(self):
...
+        """
+        # store environment to replace later
+        backup_env = os.environ
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: I don't think this method of saving/restoring `os.environ` works.  I think you need to assign into `os.environ` instead of overwriting it, or this will only affect the python level.  I wrote a test program to check this, and this shows no env changes:

```python
import os
import subprocess

os.environ = {"FOO": "bar"}
subprocess.check_call(["env"])
```

This, on the other hand, removes a variable from the output:
```python
import os
import subprocess

del os.environ["FOO"]
subprocess.check_call(["env"])
```

Was it different on the cray machine?

File: lib/spack/spack/compiler.py [hunk 14]
```diff
@@ -406,6 +406,40 @@ def c11_flag(self):
...
+        env = spack.util.environment.EnvironmentModifications()
+        env.extend(spack.schema.environment.parse(self.environment))
+        env.apply_modifications()
+
+        cc = spack.util.executable.Executable(self.cc)
+        output = cc(self.version_argument,
+                    output=str, error=str,
+                    ignore_errors=type(self.ignore_version_errors))
+
+        # restore environment
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: see note above about overwriting `os.environ`

File: lib/spack/spack/operating_systems/cray_backend.py (hunk 5 above)
Comment (tgamblin) [reply lib/spack/spack/operating_systems/cray_backend.py:88]: I'm ok with this but it needs a comment explaining why we do it.

File: lib/spack/spack/operating_systems/cray_backend.py (hunk 7 above)
Comment (tgamblin) [reply lib/spack/spack/operating_systems/cray_backend.py:114]: This also needs a comment describing what you put in the comment above.

File: lib/spack/spack/platforms/cray.py [hunk 15]
```diff
@@ -4,30 +4,30 @@
...
 
 
-def _get_modules_in_modulecmd_output(output):
-    '''Return list of valid modules parsed from modulecmd output string.'''
-    return [i for i in output.splitlines()
-            if len(i.split()) == 1]
+_craype_name_to_target_name = {
+    'x86_cascadelake': 'cascadelake',
+    'x86_naples': 'zen',
+    'x86_rome': 'zen',  # Cheating because we have the wrong modules on rzcrayz
```
Comment (tgamblin) [inline lib/spack/spack/platforms/cray.py]: What do we do about this one?  Can we fix the modules on `rzcrayz`?  I don't want this one merged to `develop`.

File: lib/spack/spack/platforms/cray.py (hunk 2 above)
Comment (tgamblin) [reply lib/spack/spack/platforms/cray.py:55]: It looks like that is merged -- can this be resolved?

File: lib/spack/spack/compilers/cce.py [hunk 16]
```diff
@@ -60,6 +71,8 @@ def c99_flag(self):
 
     @property
     def c11_flag(self):
+        if self.version >= ver('9.0'):
+            return '-std=c11'
         if self.version >= ver('8.5'):
```
Comment (tgamblin) [inline lib/spack/spack/compilers/cce.py]: Make this `elif` for maintainability -- in case the line above changes.

File: lib/spack/spack/platforms/cray.py [hunk 17]
```diff
@@ -104,9 +98,15 @@ def setup_platform_environment(cls, pkg, env):
         env.append_path("PKG_CONFIG_PATH", "/usr/lib64/pkgconfig")
         env.append_path("PKG_CONFIG_PATH", "/usr/local/lib64/pkgconfig")
 
+        if os.environ.get('CRAY_LD_LIBRARY_PATH'):
```
Comment (tgamblin) [inline lib/spack/spack/platforms/cray.py:104]: add a comment about what this var is.

File: lib/spack/spack/platforms/cray.py [hunk 18]
```diff
@@ -104,9 +98,15 @@ def setup_platform_environment(cls, pkg, env):
...
+            env.prepend_path('LD_LIBRARY_PATH',
+                             os.environ['CRAY_LD_LIBRARY_PATH'])
+
     @classmethod
     def detect(cls):
-        return os.environ.get('CRAYPE_VERSION') is not None
+        # If the cray module-based programming environment is available in
+        # MODULEPATH, then this is a cray platform.
+        return 'opt/cray' in os.environ.get('MODULEPATH', '')
```
Comment (tgamblin) [inline lib/spack/spack/platforms/cray.py:121]: add a longer docstring for detect that describes why we do it this way, and the distinction we're drawing between module- and non-module cray environments.   Basically explain that if a cray machine doesn't have this , it's linux.

File: lib/spack/spack/platforms/cray.py [hunk 19]
```diff
@@ -119,22 +119,64 @@ def _default_target_from_env(self):
...
 
     def _avail_targets(self):
         '''Return a list of available CrayPE CPU targets.'''
+
+        def modules_in_output(output):
+            """Returns a list of valid modules parsed from modulecmd output"""
+            return [i for i in re.split(r'\s\s+|\n', output)]
+
+        def target_names_from_modules(modules):
+            # Craype- module prefixes that are not valid CPU targets.
```
Comment (tgamblin) [inline lib/spack/spack/platforms/cray.py:161]: Is it better to blacklist or whitelist this?  can we do it instead by looking at which modules start with `craype-<CPU family>`?  This seems a bit nasty and potentially brittle

File: lib/spack/spack/spec.py [hunk 20]
```diff
@@ -2232,14 +2232,25 @@ def concretize(self, tests=False):
             mvar._patches_in_order_of_appearance = list(
                 t[-1] for t in ordered_hashes)
 
+        # This is awful!!!!
+        # TODO: anything better
+        # This allows us to access the package class for externals that
+        # override their prefix. We then mark the spec as concrete again after
+        # setting the external paths.
+        self._mark_concrete()
```
Comment (tgamblin) [inline lib/spack/spack/spec.py]: why is this needed?  you don't need the specs to be concrete just to get the package - is there something I'm missing here?

File: lib/spack/spack/spec.py [hunk 21]
```diff
@@ -2232,14 +2232,25 @@ def concretize(self, tests=False):
...
             if s.external_module and not s.external_path:
                 compiler = spack.compilers.compiler_for_spec(
                     s.compiler, s.architecture)
                 for mod in compiler.modules:
                     md.load_module(mod)
 
-                s.external_path = md.get_path_from_module(s.external_module)
+                try:
+                    s.external_path = s.package.external_prefix
+                except AttributeError:
```
Comment (tgamblin) [inline lib/spack/spack/spec.py]: Prefer to use `.get()` instead of catching `AttributeError`.  `AttributeError` can easily catch the wrong thing (any `.` above) and you can miss a bug entirely.

File: lib/spack/spack/test/architecture.py [hunk 22]
```diff
@@ -210,8 +209,8 @@ def test_optimization_flags_with_custom_versions(
     target = spack.architecture.Target(target_str)
     if real_version:
         monkeypatch.setattr(
-            spack.compiler.Compiler, 'cc_version', lambda x, y: real_version
-        )
+            spack.compiler.Compiler, 'compute_real_version',
+            lambda x: real_version)
```
Comment (tgamblin) [inline lib/spack/spack/test/architecture.py:213]: is the purpose of "real version" documented somewhere?  If not, it should be.  I think that can be done generically by just saying it's to make up for bad versions in `compilers.yaml`.

File: lib/spack/spack/util/module_cmd.py [hunk 23]
```diff
@@ -87,7 +87,10 @@ def get_path_args_from_module_line(line):
         words_and_symbols = line.split(lua_quote)
         path_arg = words_and_symbols[-2]
     else:
-        path_arg = line.split()[2]
+        try:
+            path_arg = line.split()[2]
+        except IndexError:
```
Comment (tgamblin) [inline lib/spack/spack/util/module_cmd.py]: use a conditional instead of an exception here.

File: var/spack/repos/builtin/packages/cray-libsci/package.py [hunk 24]
```diff
@@ -40,4 +78,4 @@ def scalapack_libs(self):
         return self.blas_libs
 
     def install(self, spec, prefix):
-        raise NoBuildError(spec)
+        NoBuildError()
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cray-libsci/package.py]: shouldn't this still raise?

File: lib/spack/spack/compiler.py (hunk 9 above)
Comment (becker33) [reply lib/spack/spack/compiler.py]: I should expand on that. Most of the methods to get version info are class methods, but this one is not (because it needs access to the compiler object to setup the modules to be able to call it properly).

File: lib/spack/spack/compiler.py (hunk 13 above)
Comment (becker33) [reply lib/spack/spack/compiler.py]: I had not checked that, in most situations the only environment Spack is going to call out to after doing this is the build environment, which will have the same modifications as part of its setup, so this hadn't affected the tests. Thanks for catching that.

File: lib/spack/spack/platforms/cray.py (hunk 19 above)
Comment (becker33) [reply lib/spack/spack/platforms/cray.py:161]: Re-worked completely

File: lib/spack/spack/test/architecture.py (hunk 22 above)
Comment (becker33) [reply lib/spack/spack/test/architecture.py:213]: I think this is adequately documented where the method is called in `architecture.py`

File: lib/spack/spack/spec.py (hunk 20 above)
Comment (becker33) [reply lib/spack/spack/spec.py]: `Spec.package` caches the value of package, and instantiates it with the spec. I think in this case we might be able to get away with it since we know we're about to mark it concrete, but I have to think through the details of that.

When I started the branch we had an assert at the top of `Spec.package`, but that's apparently outdated.

**Comment by becker33** [comment]
@tgamblin I think I've addressed all of your review comments.

File: lib/spack/spack/compiler.py [hunk 25]
```diff
@@ -406,6 +407,53 @@ def c11_flag(self):
                                       "the C11 standard",
                                       "c11_flag")
 
+    # Run the compiler to get compiler version info
+    # Use the run environment of the compiler to do so
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: These two lines seem redundant with the docstring now.

File: lib/spack/spack/spec.py (hunk 20 above)
Comment (tgamblin) [reply lib/spack/spack/spec.py]: Looks like this is gone.  marking resolved.

File: lib/spack/spack/compiler.py (hunk 13 above)
Comment (tgamblin) [reply lib/spack/spack/compiler.py]: This still doesn't work because `backup_env` is a reference, not a copy.  You want:

```python
backup_env = os.environ.copy()
```

File: lib/spack/spack/compiler.py [hunk 26]
```diff
@@ -406,6 +407,53 @@ def c11_flag(self):
...
+
+        # restore environment
+        # replacing the os.environ dictionary in python does not effect the
+        # environment variables in process space, so we have to modify the
+        # environment in place
+        for key in set(os.environ.keys()) | set(backup_env.keys()):
+            if key in backup_env:
+                os.environ[key] = backup_env[key]
+            else:
+                del os.environ[key]
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: effect -> affect

Also: Instead of this loop, just use `.copy()` above and do:

```
os.environ.clear()
os.environ.update(backup_env)
```

This makes `os.environ == backup_env` in-place (since it's some special kind of `dict`).

File: lib/spack/spack/compiler.py [hunk 27]
```diff
@@ -406,6 +407,45 @@ def c11_flag(self):
...
+        """Query the compiler for its version.
+
+        This is the "real" compiler version, regardless of what is in the
+        compilers.yaml file, which the user can change to name their compiler.
+
+        Use the runtime environment of the compiler (modules and environment
+        modifications) to enable the compiler to run properly on any platform.
+        """
+        # store environment to replace later
+        backup_env = os.environ.copy
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: Still wrong 😢 

Can you please add a test for this?

**Comment by adamjstewart** [comment]
This PR broke Spack support on Blue Waters:
```console
$ spack help
Traceback (most recent call last):
  File "/projects/eot/bbcj/stewart1/spack/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/main.py", line 767, in main
    if spack.config.get('config:debug'):
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/config.py", line 671, in get
    return config.get(path, default, scope)
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/llnl/util/lang.py", line 552, in __getattr__
    return getattr(self.instance, name)
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/llnl/util/lang.py", line 548, in instance
    self._instance = self.factory()
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/config.py", line 653, in _config
    _add_platform_scope(cfg, ConfigScope, name, path)
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/config.py", line 606, in _add_platform_scope
    platform = spack.architecture.platform().name
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/llnl/util/lang.py", line 178, in _memoized_function
    func.cache[args] = func(*args)
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/architecture.py", line 517, in platform
    return platform_cls()
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/platforms/cray.py", line 49, in __init__
    for target in self._avail_targets():
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/platforms/cray.py", line 188, in _avail_targets
    craype_targets = target_names_from_modules(craype_modules)
  File "/mnt/b/projects/eot/bbcj/stewart1/spack/lib/spack/spack/platforms/cray.py", line 164, in target_names_from_modules
    for mod in modules:
TypeError: 'NoneType' object is not iterable
```

**Comment by adamjstewart** [comment]
Hopefully this is useful debugging information:
```console
$ module avail -t craype-
/u/sciteam/stewart1/spack/share/spack/modules/cray-cnl5-interlagos:
/sw/EasyBuild/modules/all:
/sw/bw/modulefiles:
/sw/xe/modulefiles:
/usr/local/modulefiles:
/opt/cray/craype/2.5.16/modulefiles:
craype-abudhabi
craype-abudhabi-cu
craype-accel-host
craype-accel-nvidia20
craype-accel-nvidia35
craype-accel-nvidia52
craype-accel-nvidia60
craype-barcelona
craype-hugepages128K
craype-hugepages16M
craype-hugepages2M
craype-hugepages512K
craype-hugepages64M
craype-hugepages8M
craype-interlagos
craype-interlagos-cu
craype-istanbul
craype-mc12
craype-mc8
craype-network-gemini
craype-network-none
craype-shanghai
/opt/cray/modulefiles:
craype-installer/1.16.2
craype-installer/1.17.0
craype-installer/1.18.0
craype-installer/1.20.0
craype-installer/1.24.2
craype-installer/1.24.3(default)
/opt/modulefiles:
/opt/cray/gem/modulefiles:
$ ls /opt/cray/pe/craype/default/modulefiles
ls: cannot access /opt/cray/pe/craype/default/modulefiles: No such file or directory
```
The problem is the second check for `modules_from_listdir` returns None. We should check whether `available_craype_modules` is None before trying to execute it.

**Comment by adamjstewart** [comment]
P.S. Quick workaround for anyone else who encounters this:
```diff
diff --git a/lib/spack/spack/platforms/cray.py b/lib/spack/spack/platforms/cray.py
index 2fccf2fe5..b036a507f 100644
--- a/lib/spack/spack/platforms/cray.py
+++ b/lib/spack/spack/platforms/cray.py
@@ -181,7 +181,7 @@ def modules_from_listdir():
         if getattr(self, '_craype_targets', None) is None:
             strategies = [
                 lambda: modules_in_output(module('avail', '-t', 'craype-')),
-                modules_from_listdir
+                #modules_from_listdir
             ]
             for available_craype_modules in strategies:
                 craype_modules = available_craype_modules()
```

**Comment by adamjstewart** [comment]
Also, note that before this PR, I got:
```console
$ spack arch
cray-cnl5-interlagos
```
but now I only get:
```console
$ spack arch
cray-cnl5-x86_64
```
so I think this also broke microarchitecture detection. That'll probably require someone else to debug.

**Comment by adamjstewart** [comment]
This PR also broke module loading in `compilers.yaml`. In my `compilers.yaml`, I have:
```yaml
- compiler:
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    operating_system: cnl5
    target: any
    modules:
    - PrgEnv-gnu
    - gcc/5.3.0
    environment: {}
    extra_rpaths: []
    flags: {}
    spec: gcc@5.3.0
```
This used to work fine. After updating to develop, I see an error message during installation:
```
configure: error: C compiler cannot create executables
```
When I inspect the `config.log`, it complains that there is no PrgEnv-* module loaded. If I explicitly load the same PrgEnv-gnu module in my job script, things succeed, but I later see errors like:
```
/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/5.3.0/include/ia32intrin.h: In function 'unsigned int __crc32b(unsigned int, unsigned char)':
/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/5.3.0/include/ia32intrin.h:63:39: error: '__builtin_ia32_crc32qi' was not declared in this scope
   return __builtin_ia32_crc32qi (__C, __V);
                                       ^
/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/5.3.0/include/ia32intrin.h: In function 'unsigned int __crc32w(unsigned int, short unsigned int)':
/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/5.3.0/include/ia32intrin.h:70:39: error: '__builtin_ia32_crc32hi' was not declared in this scope
   return __builtin_ia32_crc32hi (__C, __V);
                                       ^
/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/5.3.0/include/ia32intrin.h: In function 'unsigned int __crc32d(unsigned int, unsigned int)':
/opt/gcc/5.3.0/snos/lib/gcc/x86_64-suse-linux/5.3.0/include/ia32intrin.h:77:39: error: '__builtin_ia32_crc32si' was not declared in this scope
   return __builtin_ia32_crc32si (__C, __V);
                                       ^
```
I've never seen this before, but I'm guessing that it's because Spack is unsetting the `LD_LIBRARY_PATH` from the module. For now I'm just going to locally revert this commit so I can build the software I need to finish my final project.

File: lib/spack/spack/platforms/cray.py [hunk 28]
```diff
@@ -119,22 +131,66 @@ def _default_target_from_env(self):
         if getattr(self, 'default', None) is None:
             bash = Executable('/bin/bash')
             output = bash(
-                '-lc', 'echo $CRAY_CPU_TARGET',
+                '--norc', '--noprofile', '-lc', 'echo $CRAY_CPU_TARGET',
```
Comment (skosukhin) [inline lib/spack/spack/platforms/cray.py:134]: @becker33 I am a bit late with this but what output do you expect from this command? The environment is clean and `bash` is called in the login mode with `--noprofile`. Doesn't this mean that the output is always empty?
